plugins/remark-lint-validate-internal-links/index.js

```diff
@@ -0,0 +1,204 @@
+      if (hasUpperCase(ref)) {
+        file.message(warnings.pathWithUpperCase, node);
+        cache[ref] = warnings.pathWithUpperCase;
+        return;
+      }
+      addWarningIfFileIsMissing({
+          file,
+          node,
+          refRoot: docsRoot,
+          ref: ref.endsWith('/') ? `${ref.slice(0, -1)}.md` : `${ref}.md`,
+          originalRef: ref,
+          warning: warnings.missingDoc
+      }) || ( cache[ref] = warnings.valid );
+      return;
+    }
+    verifyAnchor(file, node, ref);
+  }
+  
+  function verifyImages(node) {
+    const ref = node.url;
+    if (hasUpperCase(ref.split('/').pop())) {
+      file.message(warnings.pathWithUpperCase, node);
+      cache[ref] = warnings.pathWithUpperCase;
+      return;
+    }
+    addWarningIfFileIsMissing({
+      file,
+      node,
+      refRoot: staticRoot,
+      ref,
+      warning: warnings.missingImage
+    }) || ( cache[ref] = warnings.valid );
+  }  
+}
+
+function verifyAnchor(file, node, ref) {
+  let [filePath, anchor] = ref.split('#');
+  if (filePath.endsWith('/')) {
+    filePath = filePath.slice(0, -1)
+  }
+  filePath = join(projectRoot, docsRoot, `${filePath}.md`);
+  const doc = getFile({ file, node, ref, filePath });
+  if (doc) {
+    let anchorFound = 0;
+    // We got the doc now turn it into an AST.
+    const tree = fromMarkdown(doc);
+    visit(tree, "heading", compare);
+    function compare(currentNode) {
+      // Here we get the current node from AST which holds the heading value but
+      // it is the original one with spaces and upper case in the beggining.
+      // The references are with lower cases only and dashes instead of spaces
+      // to be URL friendly. So we do transform what we get as heading value in
+      // order to do proper comparison.
+      const refAnchor
+        = currentNode.children[0].value.toLowerCase().replace(' ', '-');
+      if (refAnchor === anchor) {
+        anchorFound++
+      }
+    }
+    if (anchorFound) {
+      cache[ref] = warnings.valid;
+    } else {
+      file.message(warnings.missingAnchor, node);
+      cache[ref] = warnings.missingAnchor;
+    }
+  }
+}
+
+function isExternalPageRef(ref) {
+  return ref.startsWith('http');
+}
+
+function isDocRef(ref) {
+  return !/#/g.test(ref);
+}
+
+function hasUpperCase(ref) {
+  return /[A-Z]/g.test(ref);
+}
+
+function isRelativeRef(ref) {
+  return ref.startsWith('../');
+}
+
+function isLocalAnchorRef(ref) {
+  return ref.startsWith('#');
+}
+
+function isStaticRef(ref) {
+  // Add more extensions to this list if needed
+  const staticFileExtensions = ['.json', '.js', '.yaml', '.sh', '.html'];
+  function checkIfPathHasStaticExtension(prev, curr) {
+    ref.endsWith(curr) && prev++;
+    return prev;
+  };
+  return !!staticFileExtensions.reduce(checkIfPathHasStaticExtension, 0);
+}
+
+function getFile(config) {
+  const { file, node, ref, filePath } = config;
+  let doc;
+  try {
+    doc = readFileSync(filePath);
+  } catch (err) {
+    log(err);
```

Comment:
**suggestion**: you really should check the error to see if it's actually because the file exist, or just something else.  Also, is it possible to use [fs/promises](https://nodejs.org/api/fs.html#fs_file_system) instead of blocking I/O?  It might not be possible as these things don't seem to be async, but they don't have callbacks either.